### PR TITLE
[CartPromotion][API][UI] Apply on discounted flag added

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| License         | MIT

<!--
 - Bug fixes must be submitted against the 1.10 branch
 - Features and deprecations must be submitted against the master branch
 - Make sure that the correct base branch is set

 To be sure you are not breaking any Backward Compatibilities, check the documentation:
 https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html

### CODE REVIEW

File: src/Sylius/Behat/Context/Api/Admin/ManagingPromotionsContext.php [hunk 1]
```diff
@@ -132,4 +157,30 @@ public function promotionShouldNotExistInTheRegistry(PromotionInterface $promoti
...
+        );
+    }
+
+    /**
+     * @When I add it
+     */
+    public function iAddIt(): void
+    {
+        $this->client->create();
+    }
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingPromotionsContext.php:185]: Can we put it next to the others whens?

**Review by SirDomin — APPROVED** [review]

File: tests/Api/Responses/Expected/admin/post_promotion_response.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    "usageLimit": null,
+    "used": 0,
+    "startsAt": null,
+    "endsAt": null,
+    "couponBased": false,
+    "coupons": [],
+    "rules": [],
+    "actions": [],
+    "applyOnDiscounted": false,
+    "createdAt": @string@,
```
Comment (lchrusciel) [inline tests/Api/Responses/Expected/admin/post_promotion_response.json:21]:
```suggestion
    "createdAt": @string@.isDateTime(),
```

File: src/Sylius/Bundle/PromotionBundle/Resources/config/doctrine/model/Promotion.orm.xml [hunk 3]
```diff
@@ -29,6 +29,11 @@
         <field name="couponBased" column="coupon_based" type="boolean" />
         <field name="startsAt" column="starts_at" type="datetime" nullable="true"/>
         <field name="endsAt" column="ends_at" type="datetime" nullable="true"/>
+        <field name="applyOnDiscounted" column="apply_on_discounted" type="boolean">
+            <options>
+                <option name="default">0</option>
```
Comment (GSadee) [inline src/Sylius/Bundle/PromotionBundle/Resources/config/doctrine/model/Promotion.orm.xml:34]: Shouldn't the default value be true as it is the current behaviour?

File: src/Sylius/Bundle/PromotionBundle/Resources/config/doctrine/model/Promotion.orm.xml [hunk 4]
```diff
@@ -29,6 +29,11 @@
         <field name="couponBased" column="coupon_based" type="boolean" />
         <field name="startsAt" column="starts_at" type="datetime" nullable="true"/>
         <field name="endsAt" column="ends_at" type="datetime" nullable="true"/>
+        <field name="applyOnDiscounted" column="apply_on_discounted" type="boolean">
```
Comment (GSadee) [inline src/Sylius/Bundle/PromotionBundle/Resources/config/doctrine/model/Promotion.orm.xml:32]:
```suggestion
        <field name="applicableToDiscounted" column="applicable_to_discounted" type="boolean">
```
?

File: tests/Api/Responses/Expected/admin/post_promotion_response.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    "updatedAt": @string@
```
[the remark this answers is not in the document]
Comment (AdamKasp) [reply tests/Api/Responses/Expected/admin/post_promotion_response.json:22]: it doesn't work in this way, the checker see it as a string `@string@.isDateTime()`

File: src/Sylius/Bundle/PromotionBundle/Resources/config/doctrine/model/Promotion.orm.xml [hunk 6]
```diff
@@ -29,6 +29,11 @@
         <field name="couponBased" column="coupon_based" type="boolean" />
         <field name="startsAt" column="starts_at" type="datetime" nullable="true"/>
         <field name="endsAt" column="ends_at" type="datetime" nullable="true"/>
+        <field name="applicableOnDiscounted" column="applicable_on_discounted" type="boolean">
```
Comment (GSadee) [inline src/Sylius/Bundle/PromotionBundle/Resources/config/doctrine/model/Promotion.orm.xml:32]:
```suggestion
        <field name="applicableToDiscounted" column="applicable_to_discounted" type="boolean">
```

File: features/promotion/managing_promotions/adding_promotion.feature [hunk 7]
```diff
@@ -65,3 +65,13 @@ Feature: Adding a new promotion
...
+
+    @api @ui
+    Scenario: Adding a promotion not applicable on discounted by catalog promotion items
+        When I want to create a new promotion
+        And I specify its code as "FULL_METAL_PROMOTION"
+        And I name it "Full metal promotion"
+        And I set it as not applicable on discounted by catalog promotion items
+        And I add it
+        Then I should be notified that it has been successfully created
+        And the "Full metal promotion" promotion should not be applicable on discounted items
```
Comment (GSadee) [inline features/promotion/managing_promotions/adding_promotion.feature:77]:
```suggestion
        And the "Full metal promotion" promotion should not be applicable to discounted items
```

File: src/Sylius/Bundle/PromotionBundle/Resources/translations/messages.en.yml [hunk 8]
```diff
@@ -17,6 +17,8 @@ sylius:
             actions: Actions
             add_action: Add action
             add_rule: Add rule
+            applicable_to_discounted: Applicable to discounted
```
Comment (CoderMaggie) [inline src/Sylius/Bundle/PromotionBundle/Resources/translations/messages.en.yml:20]:
```suggestion
            applies_to_discounted: Applies to already discounted order items
```

File: src/Sylius/Bundle/PromotionBundle/Resources/translations/messages.en.yml [hunk 9]
```diff
@@ -17,6 +17,8 @@ sylius:
...
+            applicable_to_discounted_details: With this option cart promotion will apply to product discounted by catalog promotion.
```
Comment (CoderMaggie) [inline src/Sylius/Bundle/PromotionBundle/Resources/translations/messages.en.yml:21]:
```suggestion
            applies_to_discounted_details: With this option the cart promotion will be applied to order items already discounted by catalog promotion(s).
```

**Review by CoderMaggie — CHANGES_REQUESTED** [review]

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Adam! :1st_place_medal:

**Pull request merged**